xapian-core/api/registry.cc

```diff
@@ -107,6 +111,65 @@ register_object(map<string, T*> & registry, const T & obj)
 }
 
 template<class T>
+static inline void
+register_object(map<string, T*> & registry1, map<string, T*> & registry2,
+		const T & obj)
+{
+    string name = obj.name();
+    if (rare(name.empty())) {
+	throw Xapian::InvalidOperationError("Unable to register object - name() method returned empty string");
+    }
+
+    string short_name= obj.name();
+    if (rare(short_name.empty())) {
+	throw Xapian::InvalidOperationError("Unable to register object - short_name() method returned empty string");
+    }
+
+    pair<typename map<string, T *>::iterator, bool> r1;
+    r1 = registry1.insert(make_pair(name, static_cast<T*>(NULL)));
+    pair<typename map<string, T *>::iterator, bool> r2;
+    r2 = registry2.insert(make_pair(short_name, static_cast<T*>(NULL)));
+
+    if (!r1.second) {
+	// Existing element with this key, so replace the pointer with NULL
+	// and delete the existing pointer.
+	//
+	// If the delete throws, this will leave a NULL entry in the map, but
+	// that won't affect behaviour as we return NULL for "not found"
+	// anyway.  The memory used will be leaked if the dtor throws, but
+	// throwing exceptions from the dtor is bad form, so that's not a big
+	// problem.
+	T * p = NULL;
+	swap(p, r1.first->second);
+	delete p;
+    }
+
+   if (!r2.second) {
+	// Existing element with this key, so replace the pointer with NULL
+	// and delete the existing pointer.
+	//
+	// If the delete throws, this will leave a NULL entry in the map, but
+	// that won't affect behaviour as we return NULL for "not found"
+	// anyway.  The memory used will be leaked if the dtor throws, but
+	// throwing exceptions from the dtor is bad form, so that's not a big
+	// problem.
+	T * p = NULL;
+	swap(p, r2.first->second);
+	delete p;
+    }
+
+    T * clone = obj.clone();
+    if (rare(!clone)) {
+	throw Xapian::InvalidOperationError("Unable to register object - clone() method returned NULL");
+    }
+
+    r1.first->second = clone;
+
+    T * clone2 = obj.clone();
+    r2.first->second = clone2;
```

Comment:
@ojwb I realize we need to call `clone()` more than just once here, otherwise valgrind complains about invalid read and writes for tests `registry2` and `double_register_leak`. The problem with calling `clone()` once and storing the pointer to the returned object to both `r1.first->second` and `r2.first->second` is that with multiple calls to `register_weighting_scheme()` (as is the case in the mentioned tests) pointer gets deleted at line 144 and it then tries to delete the same pointer at line 158 and that's where errors are reported by valgrind. A simple solution is to call clone once for both `r1` and `r2` which basically makes sure the problem that I just described never occurs.